File: ICPCTrainingCamp/ICPC/2023AsiaRegional/E.cpp
using namespace std;
#include <bits/stdc++.h>
#define ll long long
#define MAXN 405
#define ALPHABETSIZE 26
#define ff first
#define ss second
#define pii pair<long long,long long>
#define vi vector<long long>
#define vii vector<pair<long long,long long>>
ll a, b, m;
map<ll, ll> visited;
ll power(ll y)
{
    if (y == 0){
        return 1;
    }
    else if(y == 1){
        return a % m;
    }
    else if(visited.count(y) > 0){
        return visited[y];
    }
    else if(y % 2 == 0){
        ll p = power(y/2) % m;
        visited[y] = ((p * p) - 2) % m;
        return visited[y];
    }
    else{
        ll p1 = power(y/2) % m;
        ll p2 = power(y/2 + 1) % m;
        visited[y] = ((p1 * p2) - a) % m;
        return visited[y];
    }
}
int main()
{
    cin >> a >> b >> m;
    ll res = power(b);
    if(res < 0) res += m;
    cout << res << "\n";
}

File: ICPCTrainingCamp/ICPC/2023AsiaRegional/E.py
def power(y, a, m, visited):
    if y == 0:
        return 1
    elif y == 1:
        return a % m
    elif y in visited:
        return visited[y]
    elif y % 2 == 0:
        p = power(y // 2, a, m, visited) % m
        visited[y] = ((p * p) - 2) % m
        return visited[y]
    else:
        p1 = power(y // 2, a, m, visited) % m
        p2 = power(y // 2 + 1, a, m, visited) % m
        visited[y] = ((p1 * p2) - a) % m
        return visited[y]


def main():
    a, b, m = map(int, input().split())
    visited = {}
    res = power(b, a, m, visited)
    if res < 0:
        res += m
    print(res)


if __name__ == "__main__":
    main()

